rig-core/src/middlewares/components.rs

```diff
@@ -0,0 +1,252 @@
+{
+    pub fn new(inner: S, integration: T) -> Self {
+        Self { inner, integration }
+    }
+}
+
+impl<S, T, Response> Service<CompletionRequest> for AwaitApprovalLayerService<S, T>
+where
+    S: Service<CompletionRequest, Response = CompletionResponse<Response>> + Clone + Send + 'static,
+    S::Future: Send,
+    Response: Clone + 'static + Send,
+    T: HumanInTheLoop + Clone + Send + 'static,
+{
+    type Response = CompletionResponse<Response>;
+    type Error = bool;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: CompletionRequest) -> Self::Future {
+        let mut inner = self.inner.clone();
+        let await_approval_loop = self.integration.clone();
+
+        Box::pin(async move {
+            let Ok(res) = inner.call(req).await else {
+                todo!("Handle error properly");
+            };
+
+            let AssistantContent::Text(Text { text }) = res.choice.first() else {
+                todo!("Handle error properly");
+            };
+
+            if await_approval_loop.send_message(&text).await.is_err() {
+                todo!("Handle error properly");
+            }
+
+            let Ok(bool) = await_approval_loop.await_approval().await else {
+                todo!("Handle error properly");
+            };
+
+            if bool {
+                Ok(res)
+            } else {
+                todo!("Handle error properly - we should abort the pipeline here if the user wants to abort");
+            }
+        })
+    }
+}
+
+pub struct AwaitApprovalLayerWithPredicate<P, R, T> {
+    integration: T,
+    predicate: R,
+    _t: PhantomData<P>,
+}
+
+impl<D, R, S, T> Layer<S> for AwaitApprovalLayerWithPredicate<D, R, T>
+where
+    T: HumanInTheLoop + Clone,
+    D: Debug,
+    R: Fn(&D) -> Pin<Box<dyn Future<Output = bool> + Send>> + Clone + Send + 'static,
+{
+    type Service = AwaitApprovalLayerServiceWithPredicate<D, R, S, T>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        let predicate = self.predicate.clone();
+        AwaitApprovalLayerServiceWithPredicate::new(
+            inner,
+            self.integration.clone(),
+            predicate,
+            self._t,
+        )
+    }
+}
+
+pub struct AwaitApprovalLayerServiceWithPredicate<D, R, S, T> {
+    inner: S,
+    integration: T,
+    predicate: R,
+    _t: PhantomData<D>,
+}
+
+impl<D, R, S, T> AwaitApprovalLayerServiceWithPredicate<D, R, S, T>
+where
+    T: HumanInTheLoop,
+    R: Fn(&D) -> Pin<Box<dyn Future<Output = bool> + Send>> + Clone + Send + 'static,
+    D: Debug,
+{
+    pub fn new(inner: S, integration: T, predicate: R, _t: PhantomData<D>) -> Self {
+        Self {
+            inner,
+            integration,
+            predicate,
+            _t,
+        }
+    }
```

Comment:
Why do we need to pass `_t: PhantomData<D>` in the constructor? Can we not just initialize it when constructing the struct?